planning/MartenV4.md

```diff
@@ -0,0 +1,454 @@
+
+1. The events get appended, and the sproc returns an array of the new event versions
+1. Build up an array of `IEvent` wrappers that have the stream and event metadata
+1. Pass the `IEvent` wrappers into the transformation code, and get an additional `UpdateBatch` of projected document inserts
+   and updates
+1. Submit the subsequent `UpdateBatch` to the database
+1. The underlying database transaction is committed
+
+The difference here is that this will take multiple batched commands to the database within the same `NpgsqlTransaction`
+
+
+
+
+### Snapshots & "Live" Projections
+
+TODO
+
+## Async Daemon
+
+For v4, the Async Daemon needs to come in a couple different possible deployment configurations:
+
+1. "Classic" Polling-Based Mode -- no queueing, just something like the existing Async Daemon that polls for new events and runs the projections. I think this will be a near rewrite in v4, but there's some 
+   existing art in [Jasper's DurabilityAgent](http://jasperfx.github.io/documentation/durability/) that might help this along
+1. Messaging-Based Mode -- use Rabbit MQ, Azure Service Bus, Kafka, etc. to gather up events and connect with the running projections
+
+For both modes, we've got these issues to solve:
+
+1. Assigning projection ownership between active nodes, including the distribution and segmentation of projection builders
+1. Triggering a projection rebuild? Or at least a switchover?
+1. Enforcing event ordering
+
+Some general thoughts:
+
+* Try to decouple the projections away from `IDocumentSession` as a way to make the projection code more efficient
+* By skipping the document session, we could go directly at individual `IStorageOperation` for updates maybe. Allowing us to more effectively parallelize the generation of 
+  document updates
+* Generate a new Postgresql function for projected documents that combines the progress tracking by projection, adds optimistic concurrency checks,
+  and upserts an array of documents. Postgresql support for arrays makes this feasible
+* We need to support zero downtime projection rebuilds
+* For aggregations, introduce some caching on the aggregated documents as an optimization. Depending on the aggregation, this might not be super effective,
+  but would be huge for rebuilding projections
+
+
+### Polling-Based Mode
+
+Just some thoughts:
+
+* Use [advisory locks](https://hashrocket.com/blog/posts/advisory-locks-in-postgres) for [leader election](https://en.wikipedia.org/wiki/Leader_election) of the async daemon. The question becomes though, do you try to do a coarse grained "active / ready standby" split where the entire async deamon runs on one node at a time, or do you try to get more fine-grained and try to spread
+different projections or segments of projections around to different nodes?
+
+* Inactive nodes will poll to try to take over the ownership of the active daemon
+
+
+### Messaging-Based Projections
```

Comment:
**General concern for that point:** Do we want to make Marten distributed database/event store? I'm not against that, I'm just afraid that effectively we don't have now enough capacity to do it solid, production-ready solution. I think that we'd need to get more work around sustainability to be able to work not only after hours. I'm all for that idea, but I think that we need to solve the financial puzzle to be able to put more focus on that...

I think that the first thing that would be good to provide would be just a subscription mechanism similar to what Event Store has (see https://eventstore.com/docs/getting-started/reading-subscribing-events/index.html?tabs=tabid-6%2Ctabid-dotnet-client%2Ctabid-dotnet-read-event%2Ctabid-create-sub-dotnet ). For the basic mechanism, it would probably even tuned async daemon would be fine. Of course, the user would need to manage it on their own with the hosted service mechanism, but it still seems to be an achievable first step.

Reply:
I mean for us to make the polling async daemon able to run in a cluster, even if that's just by running hot on one node, and cold on others.